README.md
# lib_VBA
VBA generic functions collection

## m_VBA_Lib.bas
Generic VBA functions

## m_Access.bas
Generic VBA functions that depend of ACCESS

## m_Crypt.bas
MD5 and base64 functions

## Clipboard.bas
Text2Clipboard and Clipboard2Text

## m_Extra.bas
Extra functions to access

## Twofish.cls
Class that implement twofish

## URLAsync.cls
Class to retrieve a URL in async mode

## basModExp.bas
Module Exponential function


History.txt
v2.1 (2022-03-16)
====
Moved do github

m_VBA_Lib:
  Item: added negative index to the last (-index) ones
  File_Load: changed open mode to Binary
  
m_Access:
  fn_Table: added optional timeout
  Exec_SQL: added optional timeout
  Errors_Msg: corrected msg in loop

v2.0 (2021-03-03)
====
Update to work in VBA x64
Added Clipboard.bas
m_VBA_Lib:
  added TCPConnection

v1.2 (2016-04-13)
====
Added compatibility with VBA x64

m_Access:
  fn_Scalar: added DAO.Recordset (DAO.), removed Trusted_Connection=Yes
  fn_Table: added DAO.Recordset (DAO.)
  Exec_SQL: removed dbSeeChanges (in q.Execute)
  GetString: removed bug from empty recordsource
  Form_GUID: removed bug from function return
  added GetFixedSizeTXT, Replace_Table_Connections, ADORecordsetMemory
  
m_Crypt:
  DecodeArray64: removed bug from function return
  Encode64: removed bug from function return
  
m_VBA_Lib:
  Shell_InOut: ShellEx replace vbMinimizedNoFocu to vbHide (to not remove focus from application)
  DNS: added User or parameters
  SQL_Value: Added POSTGRES server (POSTGRES or POSTGRESQL or PG)
  UTF8_Decode: Added BOF remove
  added SQL_Format, UTF8_Encode_Escaped, Date_Between
  

v1.1.8 (2012-07-04)
======
m_VBA_Lib:
  added SetClipboard, GetClipboard
  function xx: ss
  
=======
v1.1.11 (2012-09-28)
=======
m_VBA_Lib:
  added: Process_List
  removed: DSN
  added: DSN_Generic, DSN_SQLServer,DSN_Firebird,DSN_MySQL
 
m_Access:
  fn_Scalar: added UID and PWD parameters
  Form_GUID: bug
  
m_Crypt:
  DecodeArray64: bug
  EncodeArray64: bug
  

v1.1.10 (2012-08-10)
=======
m_VBA_Lib:
  RTrimEX: bug with zero lenght text
  added PegaCEP: get Brazilian CEPs
  

v1.1.9 (2012-08-06)
======
m_VBA_Lib:
  added PhysicalDisk, DiskInfo


v1.1.8 (2011-12-06)
======
m_VBA_Lib:
  removed Option Compare Database (not present in VB6)
  Recoded File_Load: error in load of binary files


>>>>>>> .r21
v1.1.7 (2011-03-31)
======
m_VBA_Lib:
  function SQL_Value: bug in date


v1.1.6 (2011-03-10)
======
m_VBA_Lib:
  added UTF8_Decode, UTF8_Encode


v1.1.5 (2011-03-05)
======
m_VBA_Lib:
  added ODBC_Control
  

v1.1.4 (2011-02-17)
======
m_VBA_Lib:
  DSN - bug in sUser variable name
m_Crypt:
  Modified Passwd_Init - added Salt parameter
  Recoded: DecodeArray64 and EncodeArray64


v1.1.3 (2011-02-03)
======
m_VBA_Lib:
  Added End_With
  SQL_Value - date allow date and time
 

v1.1.2 (2010-12-20)
======
m_VBA_Lib:
  Modified DNS - added User/Pass parameters
  Added XML_Escape
  Added Module11
 

v1.1  (2010-08-31)
====
Added File_Exist function to m_VBA_Lib
Added Encode64, Decode64 to m_Crypt


v1.01.1  (2010-08-13)
=======
MinEx bug


v1.01  (2010-07-27)
=====
GUID renamed to Form_GUID


v1.0  (2010-07-23)
====
First release


m_Extra.bas
Attribute VB_Name = "m_Extra"
Option Explicit
Option Compare Database

#If VBA7 Then
Private Declare PtrSafe Function LoadLibrary Lib "kernel32" Alias "LoadLibraryA" (ByVal lpLibFileName As String) As Long
Private Declare PtrSafe Function FreeLibrary Lib "kernel32" (ByVal hLibModule As Long) As Long
Private Declare PtrSafe Function GetCurrentThreadId Lib "kernel32" () As Long
Private Declare PtrSafe Function StopMouseWheel Lib "MouseHook" (ByVal hwnd As Long, ByVal AccessThreadID As Long, Optional ByVal bNoSubformScroll As Boolean = False, Optional ByVal blIsGlobal As Boolean = False) As Boolean
Private Declare PtrSafe Function StartMouseWheel Lib "MouseHook" (ByVal hwnd As Long) As Boolean
#Else
Private Declare Function LoadLibrary Lib "kernel32" Alias "LoadLibraryA" (ByVal lpLibFileName As String) As Long
Private Declare Function FreeLibrary Lib "kernel32" (ByVal hLibModule As Long) As Long
Private Declare Function GetCurrentThreadId Lib "kernel32" () As Long
Private Declare Function StopMouseWheel Lib "MouseHook" (ByVal hwnd As Long, ByVal AccessThreadID As Long, Optional ByVal bNoSubformScroll As Boolean = False, Optional ByVal blIsGlobal As Boolean = False) As Boolean
Private Declare Function StartMouseWheel Lib "MouseHook" (ByVal hwnd As Long) As Boolean
#End If
Private hLib As Long

Public Function MouseWheelON() As Boolean
  MouseWheelON = StartMouseWheel(Application.hWndAccessApp)
  If hLib <> 0 Then
    hLib = FreeLibrary(hLib)
  End If
End Function


Public Function MouseWheelOFF(Optional NoSubFormScroll As Boolean = False, Optional GlobalHook As Boolean = False) As Boolean
  Dim s As String
  Dim blret As Boolean
  Dim AccessThreadID As Long
  
  On Error Resume Next
  ' Our error string
  s = "Sorry...cannot find the MouseHook.dll file" & vbCrLf
  s = s & "Please copy the MouseHook.dll file to your Windows System folder or into the same folder as this Access MDB."
  
  ' OK Try to load the DLL assuming it is in the Window System folder
  hLib = LoadLibrary("MouseHook.dll")
  If hLib = 0 Then
    ' See if the DLL is in the same folder as this MDB
    ' CurrentDB works with both A97 and A2K or higher
    hLib = LoadLibrary(CurrentDBDir() & "utils\MouseHook.dll")
    If hLib = 0 Then
      MsgBox s, vbOKOnly, "MISSING MOUSEHOOK.dll FILE"
      MouseWheelOFF = False
      Exit Function
    End If
  End If
  
  ' Get the ID for this thread
  AccessThreadID = GetCurrentThreadId()
  ' Call our MouseHook function in the MouseHook dll.
  ' Please not the Optional GlobalHook BOOLEAN parameter
  ' Several developers asked for the MouseHook to be able to work with
  ' multiple instances of Access. In order to accomodate this request I
  ' have modified the function to allow the caller to
  ' specify a thread specific(this current instance of Access only) or
  ' a global(all applications) MouseWheel Hook.
  ' Only use the GlobalHook if you will be running multiple instances of Access!
  MouseWheelOFF = StopMouseWheel(Application.hWndAccessApp, AccessThreadID, NoSubFormScroll, GlobalHook)
End Function

Public Function Twofish_Encode(Text As String, key As String) As String
  If Text = "" Or key = "" Then Exit Function
  Dim T As Twofish
  Set T = New Twofish
  Twofish_Encode = T.EncryptString(Text, key, True)
End Function


Public Function Twofish_Decode(Text_base64 As String, key As String) As String
  If Text_base64 = "" Or key = "" Then Exit Function
  Dim T As Twofish
  Set T = New Twofish
  Twofish_Decode = T.DecryptString(Text_base64, key, True)
End Function


Function IndexText(s As String)
  Dim r As String
  r = UCase(Remove_Accents(s))
  r = Replace(r, " DE ", " ")
  r = Replace(r, " DA ", " ")
  r = Replace(r, " DAS ", " ")
  r = Replace(r, " DO ", " ")
  r = Replace(r, " DOS ", " ")
  r = Replace(r, "CE", "SE")
  r = Replace(r, "CI", "SI")
  r = Replace(r, "S ", " ")
  r = Replace(r, "RR", "R")
  r = Replace(r, "SS", "S")
  r = Replace(r, "LL", "L")
  r = Replace(r, "QU", "K")
  r = Replace(r, "GN", "N")
  r = Subst_Car(r, "WCNZYH-.+\ ", "UKMSI,,,,,,")
  r = Replace(r, ",", "")
  IndexText = r
End Function



Public Function CurrentDBDir() As String
  Dim strDBPath As String, P As Long
  strDBPath = CurrentDb.Name
  P = InStrRev(strDBPath, "\")
  CurrentDBDir = Left$(strDBPath, P)
End Function



Public Function Primeira_Maiuscula(texto As String) As String
  Dim x, r$
  Static Excecao$
  'If Excecao = "" Then Excecao = Le_Configuracao("Primeira_Maiuscula_Excecao")
  For Each x In Split(texto, " ")
    If x > "" Then
      If InStr(Excecao, x & ",") > 0 Then
        r = r & LCase(x) & " "
      Else
        r = r & UCase(Left(x, 1)) & LCase(Mid(x, 2)) & " "
      End If
    End If
  Next
  Primeira_Maiuscula = RTrim(r)
End Function




m_Crypt.bas
Attribute VB_Name = "m_Crypt"
Option Compare Database
Option Explicit

'============================================
'===== MD5 ==================================
'============================================

' Derived from the RSA Data Security, Inc. MD5 Message-Digest Algorithm,
' as set out in the memo RFC1321.
'
' See the VB6 project that accompanies this sample for full code comments on how
' it works.
'
' ASP VBScript code for generating an MD5 'digest' or 'signature' of a string. The
' MD5 algorithm is one of the industry standard methods for generating digital
' signatures. It is generically known as a digest, digital signature, one-way
' encryption, hash or checksum algorithm. A common use for MD5 is for password
' encryption as it is one-way in nature, that does not mean that your passwords
' are not free from a dictionary attack.
'
' This is 'free' software with the following restrictions:
'
' You may not redistribute this code as a 'sample' or 'demo'. However, you are free
' to use the source code in your own code, but you may not claim that you created
' the sample code. It is expressly forbidden to sell or profit from this source code
' other than by the knowledge gained or the enhanced value added by your own code.
'
' Use of this software is also done so at your own risk. The code is supplied as
' is without warranty or guarantee of any kind.
'
' Should you wish to commission some derivative work based on this code provided
' here, or any consultancy work, please do not hesitate to contact us.
'
' Web Site:  http://www.frez.co.uk
' E-mail:    sales@frez.co.uk
 
Private Const BITS_TO_A_BYTE = 8
Private Const BYTES_TO_A_WORD = 4
Private Const BITS_TO_A_WORD = 32
 
Private m_lOnBits(30)
Private m_l2Power(30)
 
Private Function LShift(lValue, iShiftBits)
    If iShiftBits = 0 Then
        LShift = lValue
        Exit Function
    ElseIf iShiftBits = 31 Then
        If lValue And 1 Then
            LShift = &H80000000
        Else
            LShift = 0
        End If
        Exit Function
    ElseIf iShiftBits < 0 Or iShiftBits > 31 Then
        Err.Raise 6
    End If
 
    If (lValue And m_l2Power(31 - iShiftBits)) Then
        LShift = ((lValue And m_lOnBits(31 - (iShiftBits + 1))) * m_l2Power(iShiftBits)) Or &H80000000
    Else
        LShift = ((lValue And m_lOnBits(31 - iShiftBits)) * m_l2Power(iShiftBits))
    End If
End Function
 
Private Function RShift(lValue, iShiftBits)
    If iShiftBits = 0 Then
        RShift = lValue
        Exit Function
    ElseIf iShiftBits = 31 Then
        If lValue And &H80000000 Then
            RShift = 1
        Else
            RShift = 0
        End If
        Exit Function
    ElseIf iShiftBits < 0 Or iShiftBits > 31 Then
        Err.Raise 6
    End If
 
    RShift = (lValue And &H7FFFFFFE) \ m_l2Power(iShiftBits)
 
    If (lValue And &H80000000) Then
        RShift = (RShift Or (&H40000000 \ m_l2Power(iShiftBits - 1)))
    End If
End Function
 
Private Function RotateLeft(lValue, iShiftBits)
    RotateLeft = LShift(lValue, iShiftBits) Or RShift(lValue, (32 - iShiftBits))
End Function
 
Private Function AddUnsigned(lX, lY)
    Dim lX4
    Dim lY4
    Dim lX8
    Dim lY8
    Dim lResult
 
    lX8 = lX And &H80000000
    lY8 = lY And &H80000000
    lX4 = lX And &H40000000
    lY4 = lY And &H40000000
 
    lResult = (lX And &H3FFFFFFF) + (lY And &H3FFFFFFF)
 
    If lX4 And lY4 Then
        lResult = lResult Xor &H80000000 Xor lX8 Xor lY8
    ElseIf lX4 Or lY4 Then
        If lResult And &H40000000 Then
            lResult = lResult Xor &HC0000000 Xor lX8 Xor lY8
        Else
            lResult = lResult Xor &H40000000 Xor lX8 Xor lY8
        End If
    Else
        lResult = lResult Xor lX8 Xor lY8
    End If
 
    AddUnsigned = lResult
End Function
 
Private Function f(x, y, z)
    f = (x And y) Or ((Not x) And z)
End Function
 
Private Function G(x, y, z)
    G = (x And z) Or (y And (Not z))
End Function
 
Private Function H(x, y, z)
    H = (x Xor y Xor z)
End Function
 
Private Function i(x, y, z)
    i = (y Xor (x Or (Not z)))
End Function
 
Private Sub FF(a, b, c, D, x, s, ac)
    a = AddUnsigned(a, AddUnsigned(AddUnsigned(f(b, c, D), x), ac))
    a = RotateLeft(a, s)
    a = AddUnsigned(a, b)
End Sub
 
Private Sub GG(a, b, c, D, x, s, ac)
    a = AddUnsigned(a, AddUnsigned(AddUnsigned(G(b, c, D), x), ac))
    a = RotateLeft(a, s)
    a = AddUnsigned(a, b)
End Sub
 
Private Sub HH(a, b, c, D, x, s, ac)
    a = AddUnsigned(a, AddUnsigned(AddUnsigned(H(b, c, D), x), ac))
    a = RotateLeft(a, s)
    a = AddUnsigned(a, b)
End Sub
 
Private Sub II(a, b, c, D, x, s, ac)
    a = AddUnsigned(a, AddUnsigned(AddUnsigned(i(b, c, D), x), ac))
    a = RotateLeft(a, s)
    a = AddUnsigned(a, b)
End Sub
 
Private Function ConvertToWordArray(sMessage)
    Dim lMessageLength
    Dim lNumberOfWords
    Dim lWordArray()
    Dim lBytePosition
    Dim lByteCount
    Dim lWordCount
 
    Const MODULUS_BITS = 512
    Const CONGRUENT_BITS = 448
 
    lMessageLength = Len(sMessage)
 
    lNumberOfWords = (((lMessageLength + ((MODULUS_BITS - CONGRUENT_BITS) \ BITS_TO_A_BYTE)) \ (MODULUS_BITS \ BITS_TO_A_BYTE)) + 1) * (MODULUS_BITS \ BITS_TO_A_WORD)
    ReDim lWordArray(lNumberOfWords - 1)
 
    lBytePosition = 0
    lByteCount = 0
    Do Until lByteCount >= lMessageLength
        lWordCount = lByteCount \ BYTES_TO_A_WORD
        lBytePosition = (lByteCount Mod BYTES_TO_A_WORD) * BITS_TO_A_BYTE
        lWordArray(lWordCount) = lWordArray(lWordCount) Or LShift(Asc(Mid(sMessage, lByteCount + 1, 1)), lBytePosition)
        lByteCount = lByteCount + 1
    Loop
 
    lWordCount = lByteCount \ BYTES_TO_A_WORD
    lBytePosition = (lByteCount Mod BYTES_TO_A_WORD) * BITS_TO_A_BYTE
 
    lWordArray(lWordCount) = lWordArray(lWordCount) Or LShift(&H80, lBytePosition)
 
    lWordArray(lNumberOfWords - 2) = LShift(lMessageLength, 3)
    lWordArray(lNumberOfWords - 1) = RShift(lMessageLength, 29)
 
    ConvertToWordArray = lWordArray
End Function
 
Private Function WordToHex(lValue)
    Dim lByte
    Dim lCount
 
    For lCount = 0 To 3
        lByte = RShift(lValue, lCount * BITS_TO_A_BYTE) And m_lOnBits(BITS_TO_A_BYTE - 1)
        WordToHex = WordToHex & Right("0" & Hex(lByte), 2)
    Next
End Function
 
Public Function MD5(sMessage)
    m_lOnBits(0) = CLng(1)
    m_lOnBits(1) = CLng(3)
    m_lOnBits(2) = CLng(7)
    m_lOnBits(3) = CLng(15)
    m_lOnBits(4) = CLng(31)
    m_lOnBits(5) = CLng(63)
    m_lOnBits(6) = CLng(127)
    m_lOnBits(7) = CLng(255)
    m_lOnBits(8) = CLng(511)
    m_lOnBits(9) = CLng(1023)
    m_lOnBits(10) = CLng(2047)
    m_lOnBits(11) = CLng(4095)
    m_lOnBits(12) = CLng(8191)
    m_lOnBits(13) = CLng(16383)
    m_lOnBits(14) = CLng(32767)
    m_lOnBits(15) = CLng(65535)
    m_lOnBits(16) = CLng(131071)
    m_lOnBits(17) = CLng(262143)
    m_lOnBits(18) = CLng(524287)
    m_lOnBits(19) = CLng(1048575)
    m_lOnBits(20) = CLng(2097151)
    m_lOnBits(21) = CLng(4194303)
    m_lOnBits(22) = CLng(8388607)
    m_lOnBits(23) = CLng(16777215)
    m_lOnBits(24) = CLng(33554431)
    m_lOnBits(25) = CLng(67108863)
    m_lOnBits(26) = CLng(134217727)
    m_lOnBits(27) = CLng(268435455)
    m_lOnBits(28) = CLng(536870911)
    m_lOnBits(29) = CLng(1073741823)
    m_lOnBits(30) = CLng(2147483647)
 
    m_l2Power(0) = CLng(1)
    m_l2Power(1) = CLng(2)
    m_l2Power(2) = CLng(4)
    m_l2Power(3) = CLng(8)
    m_l2Power(4) = CLng(16)
    m_l2Power(5) = CLng(32)
    m_l2Power(6) = CLng(64)
    m_l2Power(7) = CLng(128)
    m_l2Power(8) = CLng(256)
    m_l2Power(9) = CLng(512)
    m_l2Power(10) = CLng(1024)
    m_l2Power(11) = CLng(2048)
    m_l2Power(12) = CLng(4096)
    m_l2Power(13) = CLng(8192)
    m_l2Power(14) = CLng(16384)
    m_l2Power(15) = CLng(32768)
    m_l2Power(16) = CLng(65536)
    m_l2Power(17) = CLng(131072)
    m_l2Power(18) = CLng(262144)
    m_l2Power(19) = CLng(524288)
    m_l2Power(20) = CLng(1048576)
    m_l2Power(21) = CLng(2097152)
    m_l2Power(22) = CLng(4194304)
    m_l2Power(23) = CLng(8388608)
    m_l2Power(24) = CLng(16777216)
    m_l2Power(25) = CLng(33554432)
    m_l2Power(26) = CLng(67108864)
    m_l2Power(27) = CLng(134217728)
    m_l2Power(28) = CLng(268435456)
    m_l2Power(29) = CLng(536870912)
    m_l2Power(30) = CLng(1073741824)

    
    Dim x
    Dim k
    Dim AA
    Dim bb
    Dim cc
    Dim DD
    Dim a
    Dim b
    Dim c
    Dim D
 
    Const S11 = 7
    Const S12 = 12
    Const S13 = 17
    Const S14 = 22
    Const S21 = 5
    Const S22 = 9
    Const S23 = 14
    Const S24 = 20
    Const S31 = 4
    Const S32 = 11
    Const S33 = 16
    Const S34 = 23
    Const S41 = 6
    Const S42 = 10
    Const S43 = 15
    Const S44 = 21
 
    x = ConvertToWordArray(sMessage)
 
    a = &H67452301
    b = &HEFCDAB89
    c = &H98BADCFE
    D = &H10325476
 
    For k = 0 To UBound(x) Step 16
        AA = a
        bb = b
        cc = c
        DD = D
 
        FF a, b, c, D, x(k + 0), S11, &HD76AA478
        FF D, a, b, c, x(k + 1), S12, &HE8C7B756
        FF c, D, a, b, x(k + 2), S13, &H242070DB
        FF b, c, D, a, x(k + 3), S14, &HC1BDCEEE
        FF a, b, c, D, x(k + 4), S11, &HF57C0FAF
        FF D, a, b, c, x(k + 5), S12, &H4787C62A
        FF c, D, a, b, x(k + 6), S13, &HA8304613
        FF b, c, D, a, x(k + 7), S14, &HFD469501
        FF a, b, c, D, x(k + 8), S11, &H698098D8
        FF D, a, b, c, x(k + 9), S12, &H8B44F7AF
        FF c, D, a, b, x(k + 10), S13, &HFFFF5BB1
        FF b, c, D, a, x(k + 11), S14, &H895CD7BE
        FF a, b, c, D, x(k + 12), S11, &H6B901122
        FF D, a, b, c, x(k + 13), S12, &HFD987193
        FF c, D, a, b, x(k + 14), S13, &HA679438E
        FF b, c, D, a, x(k + 15), S14, &H49B40821
 
        GG a, b, c, D, x(k + 1), S21, &HF61E2562
        GG D, a, b, c, x(k + 6), S22, &HC040B340
        GG c, D, a, b, x(k + 11), S23, &H265E5A51
        GG b, c, D, a, x(k + 0), S24, &HE9B6C7AA
        GG a, b, c, D, x(k + 5), S21, &HD62F105D
        GG D, a, b, c, x(k + 10), S22, &H2441453
        GG c, D, a, b, x(k + 15), S23, &HD8A1E681
        GG b, c, D, a, x(k + 4), S24, &HE7D3FBC8
        GG a, b, c, D, x(k + 9), S21, &H21E1CDE6
        GG D, a, b, c, x(k + 14), S22, &HC33707D6
        GG c, D, a, b, x(k + 3), S23, &HF4D50D87
        GG b, c, D, a, x(k + 8), S24, &H455A14ED
        GG a, b, c, D, x(k + 13), S21, &HA9E3E905
        GG D, a, b, c, x(k + 2), S22, &HFCEFA3F8
        GG c, D, a, b, x(k + 7), S23, &H676F02D9
        GG b, c, D, a, x(k + 12), S24, &H8D2A4C8A
 
        HH a, b, c, D, x(k + 5), S31, &HFFFA3942
        HH D, a, b, c, x(k + 8), S32, &H8771F681
        HH c, D, a, b, x(k + 11), S33, &H6D9D6122
        HH b, c, D, a, x(k + 14), S34, &HFDE5380C
        HH a, b, c, D, x(k + 1), S31, &HA4BEEA44
        HH D, a, b, c, x(k + 4), S32, &H4BDECFA9
        HH c, D, a, b, x(k + 7), S33, &HF6BB4B60
        HH b, c, D, a, x(k + 10), S34, &HBEBFBC70
        HH a, b, c, D, x(k + 13), S31, &H289B7EC6
        HH D, a, b, c, x(k + 0), S32, &HEAA127FA
        HH c, D, a, b, x(k + 3), S33, &HD4EF3085
        HH b, c, D, a, x(k + 6), S34, &H4881D05
        HH a, b, c, D, x(k + 9), S31, &HD9D4D039
        HH D, a, b, c, x(k + 12), S32, &HE6DB99E5
        HH c, D, a, b, x(k + 15), S33, &H1FA27CF8
        HH b, c, D, a, x(k + 2), S34, &HC4AC5665
 
        II a, b, c, D, x(k + 0), S41, &HF4292244
        II D, a, b, c, x(k + 7), S42, &H432AFF97
        II c, D, a, b, x(k + 14), S43, &HAB9423A7
        II b, c, D, a, x(k + 5), S44, &HFC93A039
        II a, b, c, D, x(k + 12), S41, &H655B59C3
        II D, a, b, c, x(k + 3), S42, &H8F0CCC92
        II c, D, a, b, x(k + 10), S43, &HFFEFF47D
        II b, c, D, a, x(k + 1), S44, &H85845DD1
        II a, b, c, D, x(k + 8), S41, &H6FA87E4F
        II D, a, b, c, x(k + 15), S42, &HFE2CE6E0
        II c, D, a, b, x(k + 6), S43, &HA3014314
        II b, c, D, a, x(k + 13), S44, &H4E0811A1
        II a, b, c, D, x(k + 4), S41, &HF7537E82
        II D, a, b, c, x(k + 11), S42, &HBD3AF235
        II c, D, a, b, x(k + 2), S43, &H2AD7D2BB
        II b, c, D, a, x(k + 9), S44, &HEB86D391
 
        a = AddUnsigned(a, AA)
        b = AddUnsigned(b, bb)
        c = AddUnsigned(c, cc)
        D = AddUnsigned(D, DD)
    Next
 
    MD5 = LCase(WordToHex(a) & WordToHex(b) & WordToHex(c) & WordToHex(D))
End Function


'============================================


Public Function Passwd_Init(Name As String, Optional Salt As String = "")
  Dim Silabas As String, Senha As String, m5 As String
  Dim i As Integer, j As Long, ms As Long, ma As Long
  Silabas = "babebibocacicodadedidodufafefigagegigoguhahihojajejijojukakolalelilolumamemomunaneninopaperarererirorusasesisutatitotuvavevivovuxaxuzazezizozu"
  ms = Len(Silabas) / 2
  m5 = MD5(Salt & Name)
  Senha = Mid(Silabas, (Val("&H" & Mid(m5, 1, 2)) Mod ms) * 2 + 1, 2)
  Senha = Senha & Mid(Silabas, (Val("&H" & Mid(m5, 3, 2)) Mod ms) * 2 + 1, 2)
  Senha = Senha & Format(Abs(Val("&H" & Mid(m5, 5, 4))) Mod 10000, "0000")
  Passwd_Init = Senha
End Function


Public Function Decode64(sInput As String) As String
    If sInput = "" Then Exit Function
    Decode64 = StrConv(DecodeArray64(sInput), vbUnicode)
End Function


Public Function DecodeArray64(ByVal s As String) As Byte()
' Copyright 2007: Christian d'Heureuse, Inventec Informatik AG, Switzerland, modified by Marcos Rebello
  Dim Map1(0 To 63)  As Byte, Map2(0 To 127) As Byte
  Dim c As Integer, i As Integer
  i = 0
  For c = Asc("A") To Asc("Z"): Map1(i) = c: i = i + 1: Next
  For c = Asc("a") To Asc("z"): Map1(i) = c: i = i + 1: Next
  For c = Asc("0") To Asc("9"): Map1(i) = c: i = i + 1: Next
  Map1(i) = Asc("+"): i = i + 1
  Map1(i) = Asc("/"): i = i + 1
  For i = 0 To 127: Map2(i) = 255: Next
  For i = 0 To 63: Map2(Map1(i)) = i: Next
  
  Dim IBuf() As Byte: IBuf = StrConv(s, vbFromUnicode)
  Dim ILen As Long: ILen = UBound(IBuf) + 1
  If ILen Mod 4 <> 0 Then Err.Raise vbObjectError, , "Length of Base64 encoded input string is not a multiple of 4."
  Do While ILen > 0
    If IBuf(ILen - 1) <> Asc("=") Then Exit Do
    ILen = ILen - 1
    Loop
  Dim OLen As Long: OLen = (ILen * 3) \ 4
  Dim Out() As Byte
  ReDim Out(0 To OLen - 1) As Byte
  Dim IP As Long
  Dim op As Long
  Do While IP < ILen
    Dim i0 As Byte: i0 = IBuf(IP): IP = IP + 1
    Dim i1 As Byte: i1 = IBuf(IP): IP = IP + 1
    Dim i2 As Byte: If IP < ILen Then i2 = IBuf(IP): IP = IP + 1 Else i2 = Asc("A")
    Dim i3 As Byte: If IP < ILen Then i3 = IBuf(IP): IP = IP + 1 Else i3 = Asc("A")
    If i0 > 127 Or i1 > 127 Or i2 > 127 Or i3 > 127 Then _
      Err.Raise vbObjectError, , "Illegal character in Base64 encoded data."
    Dim b0 As Byte: b0 = Map2(i0)
    Dim b1 As Byte: b1 = Map2(i1)
    Dim b2 As Byte: b2 = Map2(i2)
    Dim b3 As Byte: b3 = Map2(i3)
    If b0 > 63 Or b1 > 63 Or b2 > 63 Or b3 > 63 Then _
      Err.Raise vbObjectError, , "Illegal character in Base64 encoded data."
    Dim o0 As Byte: o0 = (b0 * 4) Or (b1 \ &H10)
    Dim o1 As Byte: o1 = ((b1 And &HF) * &H10) Or (b2 \ 4)
    Dim o2 As Byte: o2 = ((b2 And 3) * &H40) Or b3
    Out(op) = o0: op = op + 1
    If op < OLen Then Out(op) = o1: op = op + 1
    If op < OLen Then Out(op) = o2: op = op + 1
  Loop
  DecodeArray64 = Out
End Function


Public Function Encode64(ByRef sInput As String) As String
    If sInput = "" Then Exit Function
    Dim bytTemp() As Byte
    bytTemp = StrConv(sInput, vbFromUnicode)
    Encode64 = EncodeArray64(bytTemp)
End Function


Public Function EncodeArray64(InData() As Byte) As String
' Copyright 2007: Christian d'Heureuse, Inventec Informatik AG, Switzerland, modified by Marcos Rebello
  Dim Map1(0 To 63)  As Byte, Map2(0 To 127) As Byte
  Dim InLen As Long, c As Integer, i As Integer
  i = 0
  For c = Asc("A") To Asc("Z"): Map1(i) = c: i = i + 1: Next
  For c = Asc("a") To Asc("z"): Map1(i) = c: i = i + 1: Next
  For c = Asc("0") To Asc("9"): Map1(i) = c: i = i + 1: Next
  Map1(i) = Asc("+"): i = i + 1
  Map1(i) = Asc("/"): i = i + 1
  For i = 0 To 127: Map2(i) = 255: Next
  For i = 0 To 63: Map2(Map1(i)) = i: Next
   
  InLen = UBound(InData) - LBound(InData) + 1
  If InLen = 0 Then EncodeArray64 = "": Exit Function
   
  Dim ODataLen As Long: ODataLen = (InLen * 4 + 2) \ 3     ' output length without padding
  Dim OLen As Long: OLen = ((InLen + 2) \ 3) * 4           ' output length including padding
  Dim Out() As Byte
  ReDim Out(0 To OLen - 1) As Byte
  Dim ip0 As Long: ip0 = LBound(InData)
  Dim IP As Long, op As Long
  Dim i0 As Byte, i1 As Byte, i2 As Byte
  Dim o0 As Byte, o1 As Byte, o2 As Byte, o3 As Byte
  Do While IP < InLen
     i0 = InData(ip0 + IP): IP = IP + 1
     If IP < InLen Then i1 = InData(ip0 + IP): IP = IP + 1 Else i1 = 0
     If IP < InLen Then i2 = InData(ip0 + IP): IP = IP + 1 Else i2 = 0
     o0 = i0 \ 4
     o1 = ((i0 And 3) * &H10) Or (i1 \ &H10)
     o2 = ((i1 And &HF) * 4) Or (i2 \ &H40)
     o3 = i2 And &H3F
     Out(op) = Map1(o0): op = op + 1
     Out(op) = Map1(o1): op = op + 1
     Out(op) = IIf(op < ODataLen, Map1(o2), Asc("=")): op = op + 1
     Out(op) = IIf(op < ODataLen, Map1(o3), Asc("=")): op = op + 1
  Loop
  EncodeArray64 = StrConv(Out, vbUnicode)
End Function
